### PR TITLE
Add flags to provide certificate PEM string

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
Adds flags to provide the admin CA certificate and admission webhook cert/key pair as strings, as opposed to their existing file path flags.

**Special notes for your reviewer**:
Tests are for the flags only. I didn't see tests for the details of what main.go loads, or tests elsewhere that appear to set these flags and confirm the content of internal structures.

Manual debugger verification looked fine (insofar as the variables populated and didn't fatally abort), and the webhook listen returned the correct certificate. Didn't have a proper Kong setup for the CA cert part to confirm successful verification.

Ditto negative tests: the other failure cases for args don't appear to have tests in main.go, so none added for these. Manual checks look fine.

### CODE REVIEW

File: cli/ingress-controller/main.go [hunk 1]
```diff
@@ -392,14 +404,44 @@ func main() {
 				Client: kongClient,
 			},
 		}
+		var cert tls.Certificate
+		if cliConfig.AdmissionWebhookCertPath != "" && cliConfig.AdmissionWebhookCert != "" {
+			if cliConfig.AdmissionWebhookCertPath != "/admission-webhook/tls.crt" {
```
Comment (hbagdi) [inline cli/ingress-controller/main.go:409]: The code is checking CertPath with "" and then "/admission-webhook/tls.crt`, which is redundant.
You can simplify this code for readability.

File: cli/ingress-controller/main.go [hunk 2]
```diff
@@ -392,14 +404,44 @@ func main() {
...
+				glog.Fatalf("both --admission-webhook-cert-file and --admission-webhook-cert" +
+					"are set. Please remove one or the other")
+			}
+		}
+		if cliConfig.AdmissionWebhookKeyPath != "" && cliConfig.AdmissionWebhookKey != "" {
+			if cliConfig.AdmissionWebhookKeyPath != "/admission-webhook/tls.key" {
```
Comment (hbagdi) [inline cli/ingress-controller/main.go:415]: magic string of `/admission-webhook/tls.key` used in two places with this change.
Please define it as a constant in `flags` package and then use it here.

File: cli/ingress-controller/main.go [hunk 3]
```diff
@@ -392,14 +404,40 @@ func main() {
 				Client: kongClient,
 			},
 		}
+		var cert tls.Certificate
+		if cliConfig.AdmissionWebhookCertPath != "" && cliConfig.AdmissionWebhookCert != defaultAdmissionWebhookCertPath {
```
Comment (hbagdi) [inline cli/ingress-controller/main.go:408]: It seems that the comparison here are wrong. Should this not be checking if AdmissionWebhookCertPath is not equal to default value and not empty AdmissionWebhookCert?

Same applies for key.

File: cli/ingress-controller/main.go [hunk 4]
```diff
@@ -392,14 +404,40 @@ func main() {
...
+				"are set. Please remove one or the other")
+		}
+		if cliConfig.AdmissionWebhookCert != "" {
+			var err error
+			cert, err = tls.X509KeyPair([]byte(cliConfig.AdmissionWebhookCert), []byte(cliConfig.AdmissionWebhookKey))
+			if err != nil {
+				glog.Fatalf("failed to load admission webhook cert: %s", err)
+			}
+		}
+		if cliConfig.AdmissionWebhookCertPath != "" && cliConfig.AdmissionWebhookCert == "" {
```
Comment (hbagdi) [inline cli/ingress-controller/main.go:423]: The second condition is redundant. Due to earlier checks in the code, at this point in execution, you can be sure that  `cliConfig.AdmissionWebhookCert == ""`.

File: cli/ingress-controller/main.go (hunk 3 above)
Comment (rainest) [reply cli/ingress-controller/main.go]: Similar variable names :woman_facepalming:

File: cli/ingress-controller/main.go (hunk 4 above)
Comment (rainest) [reply cli/ingress-controller/main.go:423]: Superfluous but helps with reading a bit--the link back to the check to confirm that only one is set isn't immediately obvious, and without the check it looks like it will overwrite the certs, even though it doesn't. Minor quibble though, so I've removed it.

Comment (hbagdi) [reply cli/ingress-controller/main.go:423]: Did you actually remove it? It shows up in the current diff.

Another edge case here:
If a user mounts the cert on file-system and also sets the env vars, then the certs on the file-system take precedence here. Is this intentional? I'd recommend to have the certs in env var take precedence over path based cert because that's how I've seen precedence take place.

Comment (rainest) [reply cli/ingress-controller/main.go:426]: It would probably behoove me to push commits after making them.

Thinking over this again in the morning re the second comment, I remembered I added it originally. This should prefer the envvar certs always, but the default path cert complicates that.

1. The envvar cert is set, and the path cert is the default. The initial sanity checks succeed.
2. The envvar cert is successfully loaded.
3. The path cert block is evaluated. With the check, this does nothing, because the envvar setting is nonempty and must have successfully loaded the cert, so we skip this block. Without the check, the path is nonempty because of the default, so we run the block and overwrite the cert.

We can place the path block first and allow the envvar block to override any cert set in the path block, but that's complicated by the potential for either block to fail fatally. I don't think it's likely for the default to fail (if it happens, you probably have bigger problems), but it can theoretically block loading of valid configuration.

Restored the check with an explanatory comment. We could alternately handle this by placing the path block first and emit a fatal error only if `admission-webhook-cert` is empty, but IMO both options are equally awkward.

Comment (hbagdi) [reply cli/ingress-controller/main.go:426]: Good call. Resolving and proceeding with merge now.
We don't have any unit testing around this and based on the bugs and corner-cases we have discussed, we should have added tests to assert the right behavior.
Not a blocker, merging now.
